Flaky test: TestTracingGoldenData/otlp-opencensus timed out after 10m

### Component(s)

testbed

### Describe the issue you're reporting

Seeing a lot of occurrences of this test timeout in CI:
```
panic: test timed out after 10m0s
running tests:
	TestTracingGoldenData (10m0s)
	TestTracingGoldenData/opencensus-otlp (9m30s)
```

E.g. https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/9761790955/job/26943836035?pr=33856#step:7:904

Note this is different from https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/27295, https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/27295 is due to the same port being binded twice, while this one seems to be due to `testbed.(*inProcessCollector).Stop` taking too long (deadlock?)

--- a/.chloggen/correctness-traces.yaml
+++ b/.chloggen/correctness-traces.yaml
@@ -1,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: opencensusreceiver
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Check for ErrListenerClosed error before reporting multiplexer status.
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [33865]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:
+
+# If your change doesn't affect end users or the exported elements of any package,
+# you should instead start your pull request title with [chore] or use the "Skip Changelog" label.
+# Optional: The change log or logs in which this entry should be included.
+# e.g. '[user]' or '[user, api]'
+# Include 'user' if the change is relevant to end users.
+# Include 'api' if there is a change to a library API.
+# Default: '[user]'
+change_logs: []

--- a/.chloggen/correctness-traces.yaml
+++ b/.chloggen/correctness-traces.yaml
@@ -7,7 +7,7 @@ change_type: bug_fix
 component: opencensusreceiver
 
 # A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
-note: Check for ErrListenerClosed error before reporting multiplexer status.
+note: Do not report an error into resource status during receiver shutdown when the listener connection was closed.
 
 # Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
 issues: [33865]

--- a/receiver/opencensusreceiver/opencensus.go
+++ b/receiver/opencensusreceiver/opencensus.go
@@ -113,10 +113,14 @@ func (ocr *ocReceiver) Start(ctx context.Context, host component.Host) error {
 	if !hasConsumer {
 		return errors.New("cannot start receiver: no consumers were specified")
 	}
-	ocr.ln, err = net.Listen(string(ocr.cfg.NetAddr.Transport), ocr.cfg.NetAddr.Endpoint)
-	if err != nil {
-		return fmt.Errorf("failed to bind to address %q: %w", ocr.cfg.NetAddr.Endpoint, err)
+
+	if ocr.ln == nil {
+		ocr.ln, err = net.Listen(string(ocr.cfg.NetAddr.Transport), ocr.cfg.NetAddr.Endpoint)
+		if err != nil {
+			return fmt.Errorf("failed to bind to address %q: %w", ocr.cfg.NetAddr.Endpoint, err)
+		}
 	}
+
 	// Register the grpc-gateway on the HTTP server mux
 	var c context.Context
 	c, ocr.cancel = context.WithCancel(context.Background())
@@ -155,7 +159,8 @@ func (ocr *ocReceiver) Start(ctx context.Context, host component.Host) error {
 	}()
 	go func() {
 		startWG.Done()
-		if err := ocr.multiplexer.Serve(); !errors.Is(err, cmux.ErrServerClosed) && !errors.Is(err, cmux.ErrListenerClosed) && err != nil {
+		err = ocr.multiplexer.Serve()
+		if !errors.Is(err, cmux.ErrServerClosed) && !errors.Is(err, cmux.ErrListenerClosed) && err != nil {
 			ocr.settings.TelemetrySettings.ReportStatus(component.NewFatalErrorEvent(err))
 		}
 	}()

--- a/receiver/opencensusreceiver/opencensus.go
+++ b/receiver/opencensusreceiver/opencensus.go
@@ -114,11 +114,9 @@ func (ocr *ocReceiver) Start(ctx context.Context, host component.Host) error {
 		return errors.New("cannot start receiver: no consumers were specified")
 	}
 
-	if ocr.ln == nil {
-		ocr.ln, err = net.Listen(string(ocr.cfg.NetAddr.Transport), ocr.cfg.NetAddr.Endpoint)
-		if err != nil {
-			return fmt.Errorf("failed to bind to address %q: %w", ocr.cfg.NetAddr.Endpoint, err)
-		}
+	ocr.ln, err = net.Listen(string(ocr.cfg.NetAddr.Transport), ocr.cfg.NetAddr.Endpoint)
+	if err != nil {
+		return fmt.Errorf("failed to bind to address %q: %w", ocr.cfg.NetAddr.Endpoint, err)
 	}
 
 	// Register the grpc-gateway on the HTTP server mux

--- a/receiver/opencensusreceiver/opencensus.go
+++ b/receiver/opencensusreceiver/opencensus.go
@@ -155,7 +155,7 @@ func (ocr *ocReceiver) Start(ctx context.Context, host component.Host) error {
 	}()
 	go func() {
 		startWG.Done()
-		if err := ocr.multiplexer.Serve(); !errors.Is(err, cmux.ErrServerClosed) && err != nil {
+		if err := ocr.multiplexer.Serve(); !errors.Is(err, cmux.ErrServerClosed) && !errors.Is(err, cmux.ErrListenerClosed) && err != nil {
 			ocr.settings.TelemetrySettings.ReportStatus(component.NewFatalErrorEvent(err))
 		}
 	}()

--- a/receiver/opencensusreceiver/opencensus.go
+++ b/receiver/opencensusreceiver/opencensus.go
@@ -157,8 +157,7 @@ func (ocr *ocReceiver) Start(ctx context.Context, host component.Host) error {
 	}()
 	go func() {
 		startWG.Done()
-		err = ocr.multiplexer.Serve()
-		if !errors.Is(err, cmux.ErrServerClosed) && !errors.Is(err, cmux.ErrListenerClosed) && err != nil {
+		if err := ocr.multiplexer.Serve(); !errors.Is(err, cmux.ErrServerClosed) && !errors.Is(err, cmux.ErrListenerClosed) && err != nil {
 			ocr.settings.TelemetrySettings.ReportStatus(component.NewFatalErrorEvent(err))
 		}
 	}()

--- a/receiver/opencensusreceiver/opencensus_test.go
+++ b/receiver/opencensusreceiver/opencensus_test.go
@@ -355,11 +355,13 @@ func TestStartListenerClosed(t *testing.T) {
 	resp, err := client.Do(req)
 	require.NoError(t, err, "Error posting trace to grpc-gateway server: %v", err)
 
+	defer func() {
+		require.NoError(t, resp.Body.Close())
+	}()
+
 	respBytes, err := io.ReadAll(resp.Body)
 	require.NoError(t, err)
 	respStr := string(respBytes)
-
-	require.NoError(t, resp.Body.Close())
 
 	assert.Equal(t, http.StatusOK, resp.StatusCode)
 	assert.Empty(t, respStr)

--- a/receiver/opencensusreceiver/opencensus_test.go
+++ b/receiver/opencensusreceiver/opencensus_test.go
@@ -318,13 +318,13 @@ func TestStartListenerClosed(t *testing.T) {
 	}
 	ocr := newOpenCensusReceiver(cfg, sink, nil, receivertest.NewNopSettings())
 
-	err := ocr.Start(context.Background(), componenttest.NewNopHost())
+	ctx := context.Background()
+
+	// start receiver
+	err := ocr.Start(ctx, componenttest.NewNopHost())
 	require.NoError(t, err, "Failed to start trace receiver: %v", err)
 
 	url := fmt.Sprintf("http://%s/v1/trace", addr)
-
-	// Verify that CORS is not enabled by default, but that it gives a method not allowed error.
-	verifyCorsResp(t, url, "origin.com", http.StatusNotImplemented, false)
 
 	traceJSON := []byte(`
 	{
@@ -345,6 +345,7 @@ func TestStartListenerClosed(t *testing.T) {
 	   ]
 	}`)
 
+	// send request to verify listener is working
 	req, err := http.NewRequest(http.MethodPost, url, bytes.NewBuffer(traceJSON))
 	require.NoError(t, err, "Error creating trace POST request: %v", err)
 	req.Header.Set("Content-Type", "application/json")
@@ -363,9 +364,10 @@ func TestStartListenerClosed(t *testing.T) {
 	assert.Equal(t, http.StatusOK, resp.StatusCode)
 	assert.Empty(t, respStr)
 
-	// stop the listener to see if the shutdown is blocked
+	// stop the listener
 	ocr.ln.Close()
 
+	// verify trace was sent
 	got := sink.AllTraces()
 	require.Len(t, got, 1)
 	require.Equal(t, 1, got[0].ResourceSpans().Len())
@@ -373,30 +375,12 @@ func TestStartListenerClosed(t *testing.T) {
 
 	wantNode := &commonpb.Node{Identifier: &commonpb.ProcessIdentifier{HostName: "testHost"}}
 	wantResource := &resourcepb.Resource{}
-	wantSpans := []*tracepb.Span{
-		{
-			TraceId:   []byte{0x5B, 0x8E, 0xFF, 0xF7, 0x98, 0x3, 0x81, 0x3, 0xD2, 0x69, 0xB6, 0x33, 0x81, 0x3F, 0xC6, 0xC},
-			SpanId:    []byte{0xEE, 0xE1, 0x9B, 0x7E, 0xC3, 0xC1, 0xB1, 0x73},
-			Name:      &tracepb.TruncatableString{Value: "testSpan"},
-			StartTime: timestamppb.New(time.Unix(1544712660, 0).UTC()),
-			EndTime:   timestamppb.New(time.Unix(1544712661, 0).UTC()),
-			Attributes: &tracepb.Span_Attributes{
-				AttributeMap: map[string]*tracepb.AttributeValue{
-					"attr1": {
-						Value: &tracepb.AttributeValue_IntValue{IntValue: 55},
-					},
-				},
-			},
-			Status: &tracepb.Status{},
-		},
-	}
 	assert.True(t, proto.Equal(wantNode, gotNode))
 	assert.True(t, proto.Equal(wantResource, gotResource))
-	require.Len(t, wantSpans, 1)
 	require.Len(t, gotSpans, 1)
-	assert.EqualValues(t, wantSpans[0], gotSpans[0])
 
-	require.NoError(t, ocr.Shutdown(context.Background()))
+	// stop the receiver to verify it's not blocked by the closed listener
+	require.NoError(t, ocr.Shutdown(ctx))
 }
 
 func tempSocketName(t *testing.T) string {

--- a/receiver/opencensusreceiver/opencensus_test.go
+++ b/receiver/opencensusreceiver/opencensus_test.go
@@ -25,7 +25,6 @@ import (
 	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
 	resourcepb "github.com/census-instrumentation/opencensus-proto/gen-go/resource/v1"
 	tracepb "github.com/census-instrumentation/opencensus-proto/gen-go/trace/v1"
-	gatewayruntime "github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	"go.opentelemetry.io/collector/component"
@@ -303,7 +302,7 @@ func TestStartWithoutConsumersShouldFail(t *testing.T) {
 	require.Error(t, r.Start(context.Background(), componenttest.NewNopHost()))
 }
 
-func TestStart(t *testing.T) {
+func TestStartListenerClosed(t *testing.T) {
 
 	addr := testutil.GetAvailableLocalAddress(t)
 
@@ -317,17 +316,10 @@ func TestStart(t *testing.T) {
 			},
 		},
 	}
-	ln, err := net.Listen("tcp", addr)
-	require.NoError(t, err, "failed to listen on %q: %v", addr, err)
-	ocr := &ocReceiver{
-		cfg:             cfg,
-		corsOrigins:     []string{}, // Disable CORS by default.
-		gatewayMux:      gatewayruntime.NewServeMux(),
-		traceConsumer:   sink,
-		metricsConsumer: nil,
-		ln:              ln,
-		settings:        receivertest.NewNopSettings(),
-	}
+	ocr := newOpenCensusReceiver(cfg, sink, nil, receivertest.NewNopSettings())
+
+	err := ocr.Start(context.Background(), componenttest.NewNopHost())
+	require.NoError(t, err, "Failed to start trace receiver: %v", err)
 
 	url := fmt.Sprintf("http://%s/v1/trace", addr)
 
@@ -335,23 +327,23 @@ func TestStart(t *testing.T) {
 	verifyCorsResp(t, url, "origin.com", http.StatusNotImplemented, false)
 
 	traceJSON := []byte(`
-    {
-       "node":{"identifier":{"hostName":"testHost"}},
-       "spans":[
-          {
-              "traceId":"W47/95gDgQPSabYzgT/GDA==",
-              "spanId":"7uGbfsPBsXM=",
-              "name":{"value":"testSpan"},
-              "startTime":"2018-12-13T14:51:00Z",
-              "endTime":"2018-12-13T14:51:01Z",
-              "attributes": {
-                "attributeMap": {
-                  "attr1": {"intValue": "55"}
-                }
-              }
-          }
-       ]
-    }`)
+	{
+	   "node":{"identifier":{"hostName":"testHost"}},
+	   "spans":[
+	      {
+	          "traceId":"W47/95gDgQPSabYzgT/GDA==",
+	          "spanId":"7uGbfsPBsXM=",
+	          "name":{"value":"testSpan"},
+	          "startTime":"2018-12-13T14:51:00Z",
+	          "endTime":"2018-12-13T14:51:01Z",
+	          "attributes": {
+	            "attributeMap": {
+	              "attr1": {"intValue": "55"}
+	            }
+	          }
+	      }
+	   ]
+	}`)
 
 	req, err := http.NewRequest(http.MethodPost, url, bytes.NewBuffer(traceJSON))
 	require.NoError(t, err, "Error creating trace POST request: %v", err)
@@ -371,11 +363,8 @@ func TestStart(t *testing.T) {
 	assert.Equal(t, http.StatusOK, resp.StatusCode)
 	assert.Empty(t, respStr)
 
-	ln.Close()
-
-	err = ocr.Start(context.Background(), componenttest.NewNopHost())
-	require.NoError(t, err, "Failed to start trace receiver: %v", err)
-	t.Cleanup(func() { require.NoError(t, ocr.Shutdown(context.Background())) })
+	// stop the listener to see if the shutdown is blocked
+	ocr.ln.Close()
 
 	got := sink.AllTraces()
 	require.Len(t, got, 1)
@@ -406,6 +395,8 @@ func TestStart(t *testing.T) {
 	require.Len(t, wantSpans, 1)
 	require.Len(t, gotSpans, 1)
 	assert.EqualValues(t, wantSpans[0], gotSpans[0])
+
+	require.NoError(t, ocr.Shutdown(context.Background()))
 }
 
 func tempSocketName(t *testing.T) string {

--- a/receiver/opencensusreceiver/opencensus_test.go
+++ b/receiver/opencensusreceiver/opencensus_test.go
@@ -25,6 +25,7 @@ import (
 	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
 	resourcepb "github.com/census-instrumentation/opencensus-proto/gen-go/resource/v1"
 	tracepb "github.com/census-instrumentation/opencensus-proto/gen-go/trace/v1"
+	gatewayruntime "github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	"go.opentelemetry.io/collector/component"
@@ -300,6 +301,111 @@ func TestStartWithoutConsumersShouldFail(t *testing.T) {
 	require.NotNil(t, r)
 
 	require.Error(t, r.Start(context.Background(), componenttest.NewNopHost()))
+}
+
+func TestStart(t *testing.T) {
+
+	addr := testutil.GetAvailableLocalAddress(t)
+
+	// Set the buffer count to 1 to make it flush the test span immediately.
+	sink := new(consumertest.TracesSink)
+	cfg := &Config{
+		ServerConfig: configgrpc.ServerConfig{
+			NetAddr: confignet.AddrConfig{
+				Endpoint:  addr,
+				Transport: "tcp",
+			},
+		},
+	}
+	ln, err := net.Listen("tcp", addr)
+	require.NoError(t, err, "failed to listen on %q: %v", addr, err)
+	ocr := &ocReceiver{
+		cfg:             cfg,
+		corsOrigins:     []string{}, // Disable CORS by default.
+		gatewayMux:      gatewayruntime.NewServeMux(),
+		traceConsumer:   sink,
+		metricsConsumer: nil,
+		ln:              ln,
+		settings:        receivertest.NewNopSettings(),
+	}
+
+	url := fmt.Sprintf("http://%s/v1/trace", addr)
+
+	// Verify that CORS is not enabled by default, but that it gives a method not allowed error.
+	verifyCorsResp(t, url, "origin.com", http.StatusNotImplemented, false)
+
+	traceJSON := []byte(`
+    {
+       "node":{"identifier":{"hostName":"testHost"}},
+       "spans":[
+          {
+              "traceId":"W47/95gDgQPSabYzgT/GDA==",
+              "spanId":"7uGbfsPBsXM=",
+              "name":{"value":"testSpan"},
+              "startTime":"2018-12-13T14:51:00Z",
+              "endTime":"2018-12-13T14:51:01Z",
+              "attributes": {
+                "attributeMap": {
+                  "attr1": {"intValue": "55"}
+                }
+              }
+          }
+       ]
+    }`)
+
+	req, err := http.NewRequest(http.MethodPost, url, bytes.NewBuffer(traceJSON))
+	require.NoError(t, err, "Error creating trace POST request: %v", err)
+	req.Header.Set("Content-Type", "application/json")
+
+	client := &http.Client{}
+	defer client.CloseIdleConnections()
+	resp, err := client.Do(req)
+	require.NoError(t, err, "Error posting trace to grpc-gateway server: %v", err)
+
+	respBytes, err := io.ReadAll(resp.Body)
+	require.NoError(t, err)
+	respStr := string(respBytes)
+
+	require.NoError(t, resp.Body.Close())
+
+	assert.Equal(t, http.StatusOK, resp.StatusCode)
+	assert.Empty(t, respStr)
+
+	ln.Close()
+
+	err = ocr.Start(context.Background(), componenttest.NewNopHost())
+	require.NoError(t, err, "Failed to start trace receiver: %v", err)
+	t.Cleanup(func() { require.NoError(t, ocr.Shutdown(context.Background())) })
+
+	got := sink.AllTraces()
+	require.Len(t, got, 1)
+	require.Equal(t, 1, got[0].ResourceSpans().Len())
+	gotNode, gotResource, gotSpans := opencensus.ResourceSpansToOC(got[0].ResourceSpans().At(0))
+
+	wantNode := &commonpb.Node{Identifier: &commonpb.ProcessIdentifier{HostName: "testHost"}}
+	wantResource := &resourcepb.Resource{}
+	wantSpans := []*tracepb.Span{
+		{
+			TraceId:   []byte{0x5B, 0x8E, 0xFF, 0xF7, 0x98, 0x3, 0x81, 0x3, 0xD2, 0x69, 0xB6, 0x33, 0x81, 0x3F, 0xC6, 0xC},
+			SpanId:    []byte{0xEE, 0xE1, 0x9B, 0x7E, 0xC3, 0xC1, 0xB1, 0x73},
+			Name:      &tracepb.TruncatableString{Value: "testSpan"},
+			StartTime: timestamppb.New(time.Unix(1544712660, 0).UTC()),
+			EndTime:   timestamppb.New(time.Unix(1544712661, 0).UTC()),
+			Attributes: &tracepb.Span_Attributes{
+				AttributeMap: map[string]*tracepb.AttributeValue{
+					"attr1": {
+						Value: &tracepb.AttributeValue_IntValue{IntValue: 55},
+					},
+				},
+			},
+			Status: &tracepb.Status{},
+		},
+	}
+	assert.True(t, proto.Equal(wantNode, gotNode))
+	assert.True(t, proto.Equal(wantResource, gotResource))
+	require.Len(t, wantSpans, 1)
+	require.Len(t, gotSpans, 1)
+	assert.EqualValues(t, wantSpans[0], gotSpans[0])
 }
 
 func tempSocketName(t *testing.T) string {

